File: MatlabOctaveCodeSnippets/snip_systems_smothing_FFT.m
nstart=12650;%chosen after zooming the signal in impulseResponses.wav
nend=22050;%this was the chosen segment. Adjust them for your data!
if 1 %if you have impulseResponses.wav available
    [h,Fs,b]=wavread('impulseResponses.wav');
    h=double(h(nstart:nend)); %segment and cast h to double
else %use signal with few samples extracted from impulseResponses.wav
    duration = floor(nend-nstart+1); %same duration as h above
    h =[-1051, 4155, -32678, -11250, 5536, -4756, 2941, -3162];
    h = [h zeros(1,duration-length(h))]; %pad with zeros
    h = h + 10*randn(1,length(h)); %add some noise
    Fs=44100; %chosen sampling frequency
end
N=1024; %number of FFT points
M=floor(length(h)/N); %number of segments of N samples each
h=h(1:N*M); h=h(:); %make sure h is a column vector with N*M samples
xsegments=reshape(h,N,M); %segment h into M blocks
X=abs(fft(xsegments)); %obtain the magnitude for each segment
X=mean(transpose(X)); %transpose: the mean has to be over the FFTs
f=Fs/N*(0:N/2)/1000; %create abscissa in kHz. Fs/N is the bin width
plot(f,20*log10(X(1:N/2+1))),xlabel('f (kHz)'),ylabel('|H(f)| (dBW)')

File: PythonCodeSnippets/snip_systems_smothing_FFT.py
import numpy as np
import matplotlib.pyplot as plt

nstart = 12650  # chosen after zooming the signal in impulseResponses.wav
nend = 22050  # this was the chosen segment. Adjust them for your data!

# Check if impulseResponses.wav is available
if False:  # Assuming the file is available
    # Read impulse response
    h, Fs, b = readwav('impulseResponses.wav')
    # Segment and cast h to double
    h = np.double(h[nstart:nend])
else:
    # Use signal with few samples extracted from impulseResponses.wav
    duration = nend - nstart + 1  # Same duration as h above
    h = [-1051, 4155, -32678, -11250, 5536, -4756, 2941, -3162]
    h = np.hstack([h, np.zeros(duration - len(h))])  # Pad with zeros
    h = h + 10 * np.random.randn(len(h))  # Add some noise
    Fs = 44100  # Chosen sampling frequency

N = 1024  # Number of FFT points
M = len(h) // N  # Number of segments of N samples each
h = h[:N * M].reshape(N, M, order='F')  # Segment h into M blocks
X = np.abs(np.fft.fft(h, axis=0))  # Obtain the magnitude for each segment
X = np.mean(X, axis=1)  # The mean has to be over the FFTs
f = np.arange(N//2 + 1) * Fs / N / 1000  # Create abscissa in kHz

plt.plot(f, 20 * np.log10(X[:N//2 + 1]))
plt.xlabel('f (kHz)')
plt.ylabel('|H(f)| (dBW)')
plt.show()
